# Hands-On Exercise 2: Hospital Patient Data Cleaning and Analysis

## Introduction

This exercise focuses on inspecting, cleaning, and analysing a hospital patient dataset using Pandas. The analysis will involve identifying missing values, examining data types, addressing inconsistent entries, and using the cleaned dataset to answer basic questions about patient records.

## Importing Pandas

Pandas is imported to provide the tools required to work with, clean, and analyze the hospital patient dataset.

In [2]:
import pandas as pd

In [6]:
df = pd.read_csv("hospital_patient_data_cleaning_practice.csv")

In [8]:
df.head()

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
0,1745,Sani Olawale,58.0,Male,Asaba,3,2026,Specialist Consultation
1,1679,Daniel James,49.0,Male,Owerri,10,2025,Specialist Consultation
2,3291,John Ibekwe,20.0,Female,Asaba,9,2026,Routine Check
3,3489,Bola Sani,56.0,Female,Ilorin,2,2024,Routine Check
4,1943,Aisha Chukwu,50.0,Male,Lagos,11,2026,Routine Check


### Interpretation

The first five records provide an initial view of the hospital patient dataset. The dataset contains information on patient identification, patient name, age, gender, city, visit count, record year, and visit type. The displayed records show patients from different cities, with varying ages, visit counts, record years, and types of hospital visits. The dataset has been successfully loaded into the Pandas DataFrame and is ready for further structural and data-quality inspection.

In [12]:
df.shape

(3000, 8)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Patient_ID    3000 non-null   int64  
 1   Patient_Name  3000 non-null   object 
 2   Age           2955 non-null   float64
 3   Gender        3000 non-null   object 
 4   City          2881 non-null   object 
 5   Visit_Count   2959 non-null   object 
 6   Record_Year   3000 non-null   int64  
 7   Visit_Type    3000 non-null   object 
dtypes: float64(1), int64(2), object(5)
memory usage: 187.6+ KB


### Interpretation

The dataset contains 3,000 patient records and 8 variables. The `df.info()` output identifies missing values in three variables: Age has 2,955 non-null observations, City has 2,881 non-null observations, and Visit_Count has 2,959 non-null observations. This corresponds to 45 missing Age values, 119 missing City values, and 41 missing Visit_Count values. In addition, Visit_Count is stored as an `object` data type rather than a numerical type, suggesting that the variable may contain non-numerical entries that require further investigation. The remaining variables contain complete observations.

In [15]:
df.isnull().sum()

Patient_ID        0
Patient_Name      0
Age              45
Gender            0
City            119
Visit_Count      41
Record_Year       0
Visit_Type        0
dtype: int64

### Investigation of Missing Values

The missing-value assessment identified missing observations in three variables. Age has 45 missing values, City has 119 missing values, and Visit_Count has 41 missing values. The remaining five variables have no missing observations. These findings indicate that the missing values require further consideration before the dataset is used for analysis.

In [16]:
df["Visit_Count"].unique()

array(['3', '10', '9', '2', '11', '7', 'six', '1', '6', '5', nan, '12',
       '4', '8', '4 visits', 'seven', 'ten', '0'], dtype=object)

### Investigation of Visit Count

The unique values in the `Visit_Count` variable reveal several inconsistencies. Although most observations are numerical values stored as text, some entries are written as words, including `six`, `seven`, and `ten`. Another entry is recorded as `4 visits`, while some observations are missing. These inconsistencies explain why Pandas classified the variable as an `object` rather than a numerical data type. The values therefore require standardisation before `Visit_Count` can be used reliably for numerical analysis.

In [17]:
df["Visit_Count"].value_counts(dropna=False)

Visit_Count
1           256
6           255
9           253
8           245
7           245
12          242
4           241
10          240
3           234
11          231
5           224
2           221
NaN          41
4 visits     18
seven        18
ten          18
six          17
0             1
Name: count, dtype: int64

### Frequency of Inconsistent Visit Count Entries

The frequency assessment shows that most `Visit_Count` observations are represented as numerical values stored as text. However, 18 records contain `4 visits`, 18 contain `seven`, 18 contain `ten`, and 17 contain `six`. There are also 41 missing observations. These entries represent inconsistent formats rather than necessarily invalid visit counts. The inconsistent textual representations can therefore be standardised to their corresponding numerical values, while the missing observations require separate treatment.

In [18]:
df[df["Age"].isnull()].head()

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
68,1052,Joy Okafor,NaN,Male,Owerri,5,2026,Follow-up
103,3857,Ibrahim Obi,NaN,Male,Kaduna,3,2026,Routine Check
158,3965,Victor Okeke,NaN,Male,Benin City,9,2024,Specialist Consultation
180,2592,Peter Ezeh,NaN,Male,Warri,4,2025,Routine Check
206,1433,Fatima Ezeani,NaN,Female,Jos,9,2026,Follow-up


### Investigation of Missing Age Values

An inspection of records with missing Age values shows that the affected patient records still contain information for other variables, including gender, city, visit count, record year, and visit type. Therefore, the missing Age values represent incomplete observations for a single variable rather than entirely empty patient records. The missing Age values will require an appropriate treatment before analyses involving age are performed.

In [19]:
df[df["City"].isnull()].head()

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
17,1185,Chukwu Mohammed,76.0,Male,NaN,9,2025,Routine Check
92,3079,Esther Ibrahim,38.0,Male,NaN,7,2026,Routine Check
94,1680,Sarah Peter,67.0,Female,NaN,9,2025,Routine Check
114,1267,Kelechi Paul,53.0,Male,NaN,2,2025,Routine Check
119,3242,Olamide Olawale,61.0,Female,NaN,10,2026,Routine Check


### Investigation of Missing City Values

An inspection of records with missing City values shows that the affected patient records contain information for the other variables, including age, gender, visit count, record year, and visit type. Therefore, the missing City values represent incomplete information for a single variable rather than completely unusable patient records. The appropriate treatment of these missing City values will be considered during the data-cleaning stage.

In [20]:
df.duplicated().sum()

np.int64(4)

### Investigation of Duplicate Records

The duplicate-record check identified 4 completely duplicated rows in the dataset. These records contain identical values across all variables and therefore represent unnecessary duplicate observations. The duplicated records should be removed during the data-cleaning stage to prevent them from being counted more than once in subsequent analyses.

## Data Cleaning

Based on the data-quality assessment, the dataset contains duplicate records, inconsistent formats in the Visit_Count variable, and missing values. The cleaning process will address these issues systematically while preserving valid patient information.

In [21]:
df = df.drop_duplicates() ### --- dropping duplicates

In [30]:
df.duplicated().sum() ### --- verifying if duplicates still remain

np.int64(0)

In [31]:
df = df.copy()

In [37]:
df["Visit_Count"] = df["Visit_Count"].replace({
    "six": "6",
    "seven": "7",
    "ten": "10",
    "4 visits": "4"
})

In [38]:
df["Visit_Count"]                        

0        3.0
1       10.0
2        9.0
3        2.0
4       11.0
        ... 
2995    10.0
2996     4.0
2997     8.0
2998    12.0
2999     4.0
Name: Visit_Count, Length: 2996, dtype: float64

In [39]:
df["Visit_Count"].unique()

array([ 3., 10.,  9.,  2., 11.,  7.,  6.,  1.,  5., nan, 12.,  4.,  8.,
        0.])

In [40]:
df["Visit_Count"] = pd.to_numeric(df["Visit_Count"], errors = "coerce")

In [41]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2996 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Patient_ID    2996 non-null   int64  
 1   Patient_Name  2996 non-null   object 
 2   Age           2951 non-null   float64
 3   Gender        2996 non-null   object 
 4   City          2878 non-null   object 
 5   Visit_Count   2955 non-null   float64
 6   Record_Year   2996 non-null   int64  
 7   Visit_Type    2996 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 210.7+ KB


### Standardisation of Visit Count

The inconsistent entries in the `Visit_Count` variable were standardised by converting textual representations such as `six`, `seven`, `ten`, and `4 visits` to their corresponding numerical values. The variable was subsequently converted from an `object` data type to a numerical (`float64`) data type. After removing four duplicate records, the cleaned dataset contains 2,996 records, with 2,955 non-null observations for Visit_Count.

In [42]:
df["Age"].mean(), df["Age"].median()

(np.float64(49.003727549983054), 49.0)

### Treatment of Missing Age Values

The mean and median ages in the cleaned dataset are approximately 49.00 and 49.00 years, respectively. Because the two measures are essentially identical, the median age of 49 years will be used to replace the missing Age values. This provides a simple central-value treatment while preserving the existing patient records.

In [44]:
df["Age"] = df["Age"].fillna(df["Age"].median())

In [45]:
df["Age"].isnull().sum()

np.int64(0)

### Age Cleaning Result

The missing Age values were replaced using the median age of 49 years. A subsequent missing-value check returned zero, confirming that all missing Age observations have been successfully addressed.

In [46]:
df["City"].value_counts()

City
Calabar          177
Uyo              173
Warri            170
Zaria            170
Enugu            165
Kano             165
Asaba            164
Ilorin           161
Kaduna           160
Benin City       158
Abeokuta         154
Akure            150
Port Harcourt    149
Lagos            147
Abuja            143
Jos              140
Owerri           139
Ibadan           128
ABUJA             15
ibadan            15
Port-Harcourt     15
lagos             15
TEST               5
Name: count, dtype: int64

### Investigation of City Values

The City variable contains several inconsistencies in how city names are recorded. For example, Abuja appears as both `Abuja` and `ABUJA`, Ibadan appears as both `Ibadan` and `ibadan`, Lagos appears as both `Lagos` and `lagos`, and Port Harcourt appears as both `Port Harcourt` and `Port-Harcourt`. In addition, five records contain `TEST`, which does not represent a valid city entry in the dataset. These inconsistencies should be standardised before the missing City values are addressed.

In [47]:
df[df["City"] == "TEST"]

,Patient_ID,Patient_Name,Age,Gender,City,Visit_Count,Record_Year,Visit_Type
18,9005,System Test,0.0,Unknown,TEST,NaN,2024,Test
64,9001,Test Patient One,0.0,Unknown,TEST,NaN,2026,Test
355,9004,Dummy Patient,1.0,Unknown,TEST,1.0,2025,Test
1399,9003,Clinic Test Entry,0.0,Unknown,TEST,NaN,2026,Test
2461,9002,Sample Record,99.0,Unknown,TEST,0.0,2026,Test


### Investigation of Test Records

Further inspection of the five records with `TEST` as the City shows that they are test or dummy records rather than genuine patient records. The records contain indicators such as `System Test`, `Test Patient One`, `Dummy Patient`, `Clinic Test Entry`, and `Sample Record`, with `Unknown` gender and `Test` visit types. These records are therefore considered invalid patient records and should be removed from the analytical dataset before further analysis.

In [49]:
df = df[df["City"] != "TEST"].copy()

In [50]:
df["City"].value_counts()

City
Calabar          177
Uyo              173
Warri            170
Zaria            170
Enugu            165
Kano             165
Asaba            164
Ilorin           161
Kaduna           160
Benin City       158
Abeokuta         154
Akure            150
Port Harcourt    149
Lagos            147
Abuja            143
Jos              140
Owerri           139
Ibadan           128
lagos             15
ABUJA             15
ibadan            15
Port-Harcourt     15
Name: count, dtype: int64

In [51]:
df["City"] = df["City"].replace({
    "ABUJA" : "Abuja",
    "ibadan": "Ibadan",
    "lagos" : "Lagos",
    "Port-Harcourt": "Port Harcourt"
})

In [52]:
df["City"].value_counts()

City
Calabar          177
Uyo              173
Zaria            170
Warri            170
Kano             165
Enugu            165
Port Harcourt    164
Asaba            164
Lagos            162
Ilorin           161
Kaduna           160
Benin City       158
Abuja            158
Abeokuta         154
Akure            150
Jos              140
Owerri           139
Ibadan           128
ibadan            15
Name: count, dtype: int64

In [53]:
df.loc[df["City"].str.lower().str.contains("ibadan",na=False), "City"].unique()

array(['Ibadan', 'ibadan '], dtype=object)

In [54]:
df["City"] = df["City"].str.strip()

In [55]:
df["City"].value_counts()

City
Calabar          177
Uyo              173
Zaria            170
Warri            170
Kano             165
Enugu            165
Port Harcourt    164
Asaba            164
Lagos            162
Ilorin           161
Kaduna           160
Benin City       158
Abuja            158
Abeokuta         154
Akure            150
Jos              140
Owerri           139
Ibadan           128
ibadan            15
Name: count, dtype: int64

In [56]:
df["City"] = df["City"].replace({
    "ibadan": "Ibadan"
})

In [57]:
df["City"].value_counts()

City
Calabar          177
Uyo              173
Warri            170
Zaria            170
Kano             165
Enugu            165
Asaba            164
Port Harcourt    164
Lagos            162
Ilorin           161
Kaduna           160
Benin City       158
Abuja            158
Abeokuta         154
Akure            150
Ibadan           143
Jos              140
Owerri           139
Name: count, dtype: int64

### Standardisation of City Names

Inconsistent city entries were standardised to ensure that different spellings and formatting variations represented the same city. Leading and trailing whitespace was removed, and inconsistent entries such as `ABUJA`, `ibadan`, `lagos`, and `Port-Harcourt` were standardised to `Abuja`, `Ibadan`, `Lagos`, and `Port Harcourt`, respectively. The invalid `TEST` records were previously removed because they represented test or dummy patient records.

### Final City Standardisation

After removing invalid test records and standardising inconsistent city names, the City variable contains 18 valid city categories. Variations in capitalisation, punctuation, and trailing whitespace were corrected so that entries referring to the same city are represented consistently. This standardisation improves the accuracy of subsequent frequency analyses of patient locations.

In [58]:
df["City"].isnull().sum()

np.int64(118)

### Treatment of Missing City Values

After removing invalid test records and standardising the valid city names, 118 patient records still have missing City values. Because City is a categorical variable and there is insufficient information in the affected records to determine the patients' actual locations, the missing values will not be arbitrarily assigned to the most frequent city. Retaining these values as missing avoids introducing unsupported information and artificially distorting the distribution of patients across cities.

In [59]:
df["Visit_Count"].isnull().sum()

np.int64(38)

In [60]:
df["Visit_Count"].describe()

count    2953.000000
mean        6.524213
std         3.420235
min         1.000000
25%         4.000000
50%         7.000000
75%         9.000000
max        12.000000
Name: Visit_Count, dtype: float64

### Treatment of Missing Visit Count Values

The cleaned `Visit_Count` variable has a mean of approximately 6.52 visits and a median of 7 visits, with values ranging from 1 to 12 visits. Because the mean and median are relatively close, the median value of 7 visits will be used to replace the remaining missing Visit_Count values. This preserves a realistic whole-number value for the count variable.

In [62]:
df["Visit_Count"] = df["Visit_Count"].fillna(df["Visit_Count"].median())

In [63]:
df["Visit_Count"].isnull().sum()

np.int64(0)

### Visit Count Cleaning Result

The remaining missing values in `Visit_Count` were replaced using the median value of 7 visits. A subsequent missing-value check returned zero, confirming that all missing Visit_Count observations have been addressed. The variable is now numeric and consistently formatted, making it suitable for subsequent quantitative analysis.

In [64]:
df.isnull().sum()

Patient_ID        0
Patient_Name      0
Age               0
Gender            0
City            118
Visit_Count       0
Record_Year       0
Visit_Type        0
dtype: int64

In [66]:
df.duplicated().sum()

np.int64(0)

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2991 entries, 0 to 2999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Patient_ID    2991 non-null   int64  
 1   Patient_Name  2991 non-null   object 
 2   Age           2991 non-null   float64
 3   Gender        2991 non-null   object 
 4   City          2873 non-null   object 
 5   Visit_Count   2991 non-null   float64
 6   Record_Year   2991 non-null   int64  
 7   Visit_Type    2991 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 210.3+ KB


### Final Data Quality Check

The final structural assessment shows that the cleaned dataset contains 2,991 patient records and 8 variables. The Age and Visit_Count variables contain complete observations following the imputation procedures, while 118 City values remain missing because their actual locations could not be determined from the available data. The Visit_Count variable is now stored as a numerical data type, and the invalid test records and duplicate observations have been removed. The dataset is therefore ready for the subsequent analysis.

In [68]:
df.isnull().sum()

Patient_ID        0
Patient_Name      0
Age               0
Gender            0
City            118
Visit_Count       0
Record_Year       0
Visit_Type        0
dtype: int64

In [69]:
df.duplicated().sum()

np.int64(0)

In [70]:
df.isnull().sum()

Patient_ID        0
Patient_Name      0
Age               0
Gender            0
City            118
Visit_Count       0
Record_Year       0
Visit_Type        0
dtype: int64

In [71]:
df.duplicated().sum()

np.int64(0)

## Completion of Data Cleaning

The data-cleaning process is complete. Duplicate records and invalid test entries were removed, inconsistent values in the `Visit_Count` and `City` variables were standardised, and missing Age and Visit_Count values were addressed using median imputation. The final dataset contains 2,991 patient records and 8 variables. A final quality check confirmed that no duplicate records remain and that all variables except City contain complete observations. The cleaned dataset is therefore suitable for further analysis.

In [72]:
df.describe()

,Patient_ID,Age,Visit_Count,Record_Year
count,2991.000000,2991.000000,2991.000000,2991.000000
mean,2501.595119,49.052156,6.530257,2025.239050
std,866.753674,17.903923,3.398849,0.763743
min,1001.000000,18.000000,1.000000,2024.000000
25%,1750.500000,34.000000,4.000000,2025.000000
50%,2503.000000,49.000000,7.000000,2025.000000
75%,3252.500000,65.000000,9.000000,2026.000000
max,4000.000000,80.000000,12.000000,2026.000000


### Descriptive Statistics

The descriptive statistics show that the cleaned dataset contains 2,991 patient records. The average patient age is approximately 49.05 years, with ages ranging from 18 to 80 years. Patients have an average of approximately 6.53 visits, with visit counts ranging from 1 to 12 and a median of 7 visits. The records cover the years 2024 to 2026, with 2025 being the median record year.

In [73]:
df["City"] = df["City"].fillna("Unknown")

In [74]:
df["City"].isnull().sum()

np.int64(0)

### Treatment of Missing City Values

The 118 missing values in the City variable were replaced with `Unknown` using the Pandas `fillna()` method. A subsequent missing-value check returned zero, confirming that the City variable no longer contains missing values. This approach preserves the patient records while clearly identifying observations for which the city information was unavailable.

In [77]:
df["Patient_Category"] = df["Visit_Count"].apply(
    lambda x: "Frequent Patient" if x >= 5 else "Occasional Patient"
)

In [78]:
df["Patient_Category"].value_counts()

Patient_Category
Frequent Patient      2024
Occasional Patient     967
Name: count, dtype: int64

### Patient Classification

Based on the visit-count classification rule, 2,024 patients are classified as Frequent Patients because they have five or more visits, while 967 patients are classified as Occasional Patients because they have fewer than five visits. Frequent Patients therefore represent the larger group in the cleaned dataset.

In [79]:
df["City"].value_counts()

City
Calabar          177
Uyo              173
Warri            170
Zaria            170
Kano             165
Enugu            165
Asaba            164
Port Harcourt    164
Lagos            162
Ilorin           161
Kaduna           160
Benin City       158
Abuja            158
Abeokuta         154
Akure            150
Ibadan           143
Jos              140
Owerri           139
Unknown          118
Name: count, dtype: int64

### Number of Patients by City

The distribution of patients by city shows that Calabar has the highest number of recorded patients with 177, followed by Uyo with 173 and Warri with 170. Owerri has the lowest number among the named cities with 139 patients. In addition, 118 patient records have an `Unknown` city because their city information was unavailable.

In [80]:
(df["Patient_Category"] == "Frequent Patient").sum()

np.int64(2024)

### Number of Frequent Patients

There are 2,024 Frequent Patients in the cleaned dataset. These patients have recorded visit counts of five or more, according to the classification rule specified in the exercise.

In [81]:
df["Visit_Count"].mean()

np.float64(6.530257438983617)

### Average Patient Visit Count

The average patient visit count in the cleaned dataset is approximately 6.53 visits. This indicates that, on average, each patient record represents approximately six to seven visits to the outpatient department.

In [82]:
original_df = pd.read_csv("hospital_patient_data_cleaning_practice.csv")

In [83]:
original_df.isnull().any(axis=1).sum()

np.int64(204)

### Records with Missing Information Before Cleaning

Before the cleaning process, 204 patient records contained at least one missing value. Although there were 205 missing cells across the dataset, some missing values occurred within the same patient record. Therefore, 204 represents the number of unique records affected by missing information before cleaning.

In [84]:
df.shape[0]

2991

### Records Remaining After Cleaning

After the data-cleaning process, 2,991 patient records remain in the dataset. This represents the original 3,000 records minus 4 duplicate records and 5 invalid test records that were removed during cleaning.

In [85]:
df.groupby("Patient_Category")["Visit_Count"].agg(["count", "mean"])

,count,mean
Patient_Category,,
Frequent Patient,2024,8.449111
Occasional Patient,967,2.513961


| **Patient Category** | **Number of Patients** | **Average Visits** |
| -------------------- | ---------------------: | -----------------: |
| Frequent Patient | 2,024 | 8.45 |
| Occasional Patient | 967 | 2.51 |

### Interpretation

The summary shows that Frequent Patients make up the larger group, with 2,024 patients compared with 967 Occasional Patients. Frequent Patients recorded an average of 8.45 visits, substantially higher than the 2.51 average visits among Occasional Patients, indicating a clear difference in outpatient utilisation between the two patient categories.